nvflare/app_common/workflows/statistics_controller.py

```diff
@@ -0,0 +1,286 @@
+                fl_ctx=fl_ctx,
+            )
+        elif rc in [
+            ReturnCode.EXECUTION_RESULT_ERROR,
+            ReturnCode.TASK_DATA_FILTER_ERROR,
+            ReturnCode.TASK_RESULT_FILTER_ERROR,
+        ]:
+
+            self.system_panic("Execution result is not a shareable. statistics controller is exiting.", fl_ctx=fl_ctx)
+
+        # Cleanup task result
+        client_task.result = None
+
+    def post_fn(self, task_name: str, fl_ctx: FLContext):
+
+        self.log_info(fl_ctx, "save statistics result to persistence store")
+        ds_stats = self._combine_all_metrics()
+
+        writer: JsonStatsFileWriter = fl_ctx.get_engine().get_component(self.writer_id)
+        writer.save(ds_stats, overwrite_existing=True, fl_ctx=fl_ctx)
+
+    def _combine_all_metrics(self):
+        result = {}
+        filtered_client_metrics = [metric for metric in self.client_metrics if metric in self.metric_configs]
+        filtered_global_metrics = [metric for metric in self.global_metrics if metric in self.metric_configs]
+
+        for metric in filtered_client_metrics:
+            for client in self.client_metrics[metric]:
+                for ds in self.client_metrics[metric][client]:
+                    client_dataset = f"{client}-{ds}"
+                    for feature_name in self.client_metrics[metric][client][ds]:
+                        if feature_name not in result:
+                            result[feature_name] = {}
+                        if metric not in result[feature_name]:
+                            result[feature_name][metric] = {}
+
+                        if metric == StC.STATS_HISTOGRAM:
+                            hist: Histogram = self.client_metrics[metric][client][ds][feature_name]
+                            result[feature_name][metric][client_dataset] = hist.bins
+                        else:
+                            result[feature_name][metric][client_dataset] = self.client_metrics[metric][client][ds][
+                                feature_name
+                            ]
+
+        for metric in filtered_global_metrics:
+            for ds in self.global_metrics[metric]:
+                global_dataset = f"{StC.GLOBAL}-{ds}"
+                for feature_name in self.global_metrics[metric][ds]:
+                    if metric == StC.STATS_HISTOGRAM:
+                        hist: Histogram = self.global_metrics[metric][ds][feature_name]
+                        result[feature_name][metric][global_dataset] = hist.bins
+                    else:
+                        result[feature_name][metric].update(
+                            {global_dataset: self.global_metrics[metric][ds][feature_name]}
+                        )
+
+        return result
+
+    def _get_target_metrics(self, ordered_metrics) -> List[MetricConfig]:
+        # make sure the execution order of the metrics calculation
+        targets = []
+        if self.metric_configs:
```

Comment:
Can `self.metric_configs` be None or empty dict?
If it can not, suggest adding check in ``__init__``.